presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,541 @@
+ * values.
+ *
+ * </pre>
+ */
+
+public class GroupedTypedHistogram
+        implements TypedHistogram
+{
+    static final float MAX_FILL_RATIO = 0.5f;
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTypedHistogram.class).instanceSize();
+    private static final int EMPTY_BUCKET = -1;
+    private static final int NULL = -1;
+    // JMH work indicated this was superior to linear or sum_of_count
+    private static final ProbeType PROBE_TYPE = ProbeType.SUM_OF_SQUARE;
+    private final int bucketId;
+
+    private final Type type;
+    private final BlockBuilder values;
+    private final BucketNodeFactory bucketNodeFactory;
+    //** these parallel arrays represent a node in the hash table
+    private final LongBigArray counts;
+    // need to store the groupId for a node for when we are doing value comparisons in hash lookups
+    private final LongBigArray groupIds;
+    // array of nodePointers (index in counts, valuePositions)
+    private final IntBigArray nextPointers;
+    // since we store histogram values in a hash, two histograms may have the same position (each unique value should have only one position in the internal
+    // BlockBuilder of values; not that we extract a subset of this when constructing per-group-id histograms)
+    private final IntBigArray valuePositions;
+    private final LongBigArray valueHashList = new LongBigArray();
+    //** end per-node arrays
+
+    // per groupId, we have a pointer to the first in the list of nodes for this group
+    private final LongBigArray headPointers;
+
+    private IntBigArray buckets;
+    private int nextNodePointer;
+    private int mask;
+    private int bucketCount;
+    private int maxFill;
+    // at most one thread uses this object at one time, so this must be set to the group being operated on
+    private long currentGroupId = -1;
+    private long numberOfGroups = 1;
+    //
+    private final ValueStore valueStore;
+
+    public GroupedTypedHistogram(Type type, int expectedCount)
+    {
+        checkArgument(expectedCount > 0, "expectedSize must be greater than zero");
+        this.type = type;
+        this.bucketId = expectedCount;
+        this.bucketCount = computeBucketCount(expectedCount, MAX_FILL_RATIO);
+        this.mask = bucketCount - 1;
+        this.maxFill = calculateMaxFill(bucketCount, MAX_FILL_RATIO);
+        this.values = type.createBlockBuilder(null, computeBucketCount(expectedCount, GroupedTypedHistogram.MAX_FILL_RATIO));
+        // buckets and node-arrays (bucket "points" to a node, so 1:1 relationship)
+        buckets = new IntBigArray(-1);
+        buckets.ensureCapacity(bucketCount);
+        counts = new LongBigArray();
+        valuePositions = new IntBigArray();
+        nextPointers = new IntBigArray(NULL);
+        groupIds = new LongBigArray(-1);
+        //bucket is the same as node
```

Comment:
it took me a while to wrap my head around the parallel arrays being indexed the same as buckets. probably i can remove this

because in closed hashing, a bucket is either just a pointer, or a head node with further pointers.

so i was basically stating "this is open address hashing, where each bucket is a node"

i'll remove

Reply:
done